src/main/java/edu/harvard/iq/dataverse/engine/command/impl/RegisterOIDCUserCommand.java

```
@@ -0,0 +1,166 @@
+
+    public RegisterOIDCUserCommand(DataverseRequest aRequest, String bearerToken, UserDTO userDTO) {
+        super(aRequest, (DvObject) null);
+        this.bearerToken = bearerToken;
+        this.userDTO = userDTO;
+    }
+
+    @Override
+    protected void executeImpl(CommandContext ctxt) throws CommandException {
+        try {
+            OAuth2UserRecord oAuth2UserRecord = ctxt.authentication().verifyOIDCBearerTokenAndGetOAuth2UserRecord(bearerToken);
+            UserRecordIdentifier userRecordIdentifier = oAuth2UserRecord.getUserRecordIdentifier();
+
+            if (ctxt.authentication().lookupUser(userRecordIdentifier) != null) {
+                throw new IllegalCommandException(BundleUtil.getStringFromBundle("registerOidcUserCommand.errors.userAlreadyRegisteredWithToken"), this);
+            }
+
+            boolean provideMissingClaimsEnabled = FeatureFlags.API_BEARER_AUTH_PROVIDE_MISSING_CLAIMS.enabled();
+
+            updateUserDTO(oAuth2UserRecord, provideMissingClaimsEnabled);
+
+            AuthenticatedUserDisplayInfo userDisplayInfo = new AuthenticatedUserDisplayInfo(
+                    userDTO.getFirstName(),
+                    userDTO.getLastName(),
+                    userDTO.getEmailAddress(),
+                    userDTO.getAffiliation() != null ? userDTO.getAffiliation() : "",
+                    userDTO.getPosition() != null ? userDTO.getPosition() : ""
+            );
+
+            validateUserFields(ctxt, provideMissingClaimsEnabled);
+
+            ctxt.authentication().createAuthenticatedUser(userRecordIdentifier, userDTO.getUsername(), userDisplayInfo, true);
+
+        } catch (AuthorizationException ex) {
+            throw new PermissionException(ex.getMessage(), this, null, null, true);
+        }
+    }
+
+    private void updateUserDTO(OAuth2UserRecord oAuth2UserRecord, boolean provideMissingClaimsEnabled) throws InvalidFieldsCommandException {
+        if (provideMissingClaimsEnabled) {
```

So the check that claims and json match is only done when the provideMissing is enabled? COuld/should it be checked for the other case as well? I.e. if you're not allowed to provide missing info but do should you also get a warning (at least if it is different?)? (Or have I missed that that is happening somewhere?)

---

I've decided to throw an error instead of a warning to simplify the process and prevent the user from being registered with unexpected values. We stop the flow if we detect that the user sends claims in the JSON request and the feature flag is disabled. Let me know what you think.

---

Sounds good - now regardless of the flag if you send values that won't be used, you get an error.